Add python 3.13, Remove python 3.8

sneakers-the-rat opened the pull request:
Fix: https://github.com/linkml/linkml/issues/2357

This is a big one!!!!! we can change all the `Dict[str,str]` annotations to `dict[str,str]` finally. 

I'm going to use ruff and `pyupdate` rules to make that tractable, will do the shift in several commits so the changes can be inspected. 

opening as a draft to run on CI as i go

sierra-moxon commented on .github/workflows/main.yaml:
```diff
@@ -39,15 +39,15 @@ jobs:
         exclude:
           # Test on Windows with only the oldest and newest Python versions
-          - os: windows-latest
-            python-version: "3.9"
```
this should stay, right (and add 13)?

sneakers-the-rat replied:
yes yes, just bump up the versions - these are the intermediate versions excluded from the windows test matrix, so the lowest version we're testing on windows is 3.9 (i.e. no longer exclude it), and yes we also don't exclude 3.13 so are also testing on it (when `3.13` is added to the supported versions by the setup-python action)

sneakers-the-rat commented:
Alright looks like `3.13` isn't enabled in the github actions `setup-python` yet, so this will remain a draft until that happens :)

sneakers-the-rat commented:
alright, stepped away for most of the day to do something else, but that looks like tests passing on my machine. 

I fixed pydanticgen and pythongen to also use `list` and `dict` rather than List and Dict, and then went through and found all the places where pythongen was failing and added `pytest.mark`s and regenerated snapshots that way (`python -m pytest -m pythongen --regenerate-snapshots --with-slow`) so i wouldn't get the usual spurious OWL/RDF ordering diffs (and also because it's really nice to be able to run tests limited by a specific generator so you aren't just waiting around all day when you know you didn't change anything except in one generator!)

I'm sort of (?) alarmed (?) that sqlalchemygen didn't cause any errors. I'm not an sqlalchemy or sql expert, but i thought that it would be possible to do `list[T]` in sqlalchemy to declare a one to many relationship. if not then we're good. if that is the case then sqlalchemygen is missing tests

vincentkelleher commented:
Am I the only one getting an issue with the `typing.io` import from 'pyshexc/parser/ShExDocLexer.py` ?

![image](https://github.com/user-attachments/assets/57749a29-ce5f-40c5-9729-49ca6395be52)

sneakers-the-rat commented:
> Am I the only one getting an issue with the `typing.io` import from 'pyshexc/parser/ShExDocLexer.py` ?

Ya I fixed that but an updated package was never released: https://github.com/shexSpec/grammar-python-antlr/pull/15

sierra-moxon commented:
@sneakers-the-rat - can you please fix the failing tests?

sierra-moxon commented:
@ptgolden - if you have some cycles, getting these tests fixed and merged so that the downstream PR can be completed would be wonderful.

ptgolden commented:
Happy to give it a shot, but it seems the logs from the failing tests have expired. I'm not a GitHub Actions expert-- is there a way to manually retrigger them?

ptgolden commented:
I also have a hunch this might be a chicken and egg problem. If `linkml-runtime` does not support 3.13 until https://github.com/linkml/linkml-runtime/pull/345 is merged, then these tests will fail until that PR is merged. But that PR relies on this one being merged. (Unless I'm missing something)

sierra-moxon commented:
Lets try to get the conflicts fixed and if we can isolate the test failures to just the circular dependency issue, then we can go from there on how to get this done.

sierra-moxon commented:
@sneakers-the-rat - next time, would this kind of change that may require multiple developer contributions warrant a branch instead of a fork in your opinion?

sneakers-the-rat commented:
> I also have a hunch this might be a chicken and egg problem. If linkml-runtime does not support 3.13 until https://github.com/linkml/linkml-runtime/pull/345 is merged, then these tests will fail until that PR is merged. But that PR relies on this one being merged. (Unless I'm missing something)

That's correct. I've been rebasing both since October waiting on review. They need to be merged and deployed at the same time. I can do it again tho

sneakers-the-rat commented:
so we really need a new release of `pyshexc` to be cut after https://github.com/shexSpec/grammar-python-antlr/pull/15 , that's another dependency that needs to be updated before 3.13 can work. that's one of ours right?

sneakers-the-rat commented:
ok checked that remaining test failures are from the removed `dataclasses` [monkeypatch](https://github.com/linkml/linkml-runtime/blob/main/linkml_runtime/utils/dataclass_extensions_376.py) that formerly allowed all `**kwargs`, and since the current `linkml-runtime` imports that monkeypatch (or it is imported at some point during the tests) the tests fail even if this PR removes the application of the monkeypatch in generated python modules. sort of illustrates why runtime side effects, and especially monkeypatching stdlib modules using private stdlib methods is fragile.

confirmed locally that this PR passes tests against https://github.com/linkml/linkml-runtime/pull/345/commits/0846eb0ebc054cab254f4cbca5d51c0de2fea9dc , and will work on a way of running the `linkml-runtime` upstream tests against a specific branch for cases like this where we need to do paired PRs

amc-corey-cox commented:
@sneakers-the-rat Is this just stuck on https://github.com/linkml/linkml-runtime/pull/345 ?

sneakers-the-rat commented:
@amc-corey-cox the blocker at this point is really https://github.com/shexSpec/grammar-python-antlr/pull/15 not having a new release that includes the fix that is already merged for the `typing.re` import. i'm just going to add an exception handler to that import and say "shex generation is not supported in python 3.13 take it up with the shexc repo"

sneakers-the-rat commented:
ok we yet again have battled through the mysterious wilds of our dependency tree to reach a point where things are running again. would be great to be able to merge this before things shift again!

explanation of last commits:
- got tired of waiting for `pyshexc` to release a new version, just monkeypatched `typing.io` instead. this appears to only be used in tests (which is news to me, if i knew that i would have done this from the beginning). I added a reminder in the monkeypatching function that will let us know to remove that monkeypatch if `pyshexc` ever receives and update along with instructions.
- marked a notebook test to skip until `pyshexc` receives and update. we can't monkeypatch the env that the notebook runs in without ruining the notebook, and since those are user-facing that seems bad to do. same as above, instructions in the error say to remove that skip condition as well.
- an old matplotlib version in the lockfile was pinning `numpy<2` which was causing [this to happen](https://github.com/linkml/linkml-runtime/actions/runs/14096197716/job/39483924392). I am formally retired from dealing with poetry problems in favor of using a less obnoxious package manager, but it seems like we want to update the poetry CI action to do `poetry update` rather than just `poetry lock`

sierra-moxon commented:
I pulled this PR into a branch: https://github.com/linkml/linkml/pull/2589 to trivially fix the linting issue.

amc-corey-cox commented:
I wanted to give a quick update on this PR. Sierra made a branch to fix that trivial linting issue and we are making a new release of linkml-runtime. We want to wait on the linkml-runtime release until after the weekend so we'll be holding other merges until after the weekend as well. As soon as we have the new release, we'll bump the linkml-runtime version and get these changes merged in as well.

Thank you to everyone who has put in the effort on this especially all of the work by @sneakers-the-rat. We'll get this merged early next week.

sneakers-the-rat commented:
kool. merged that in here as well. either/or, they're both identical and #2589 links back to this one. i am pretty sure i turned 'allow edits by maintainers' on? sorry you needed to make another pr to fix that

dalito commented:
Merged in #2605
